src/main/java/org/candlepin/model/OwnerProductCurator.java

```diff
@@ -289,6 +290,68 @@ public CandlepinQuery<Product> getProductsByIdsUsingOwnerKey(String ownerKey,
         return this.cpQueryFactory.<Product>buildQuery();
     }
 
+    /**
+     * Builds a query for fetching the system purpose attributes mapped to the given owner's products
+     *  Will only return attributes for products that are related to unexpired pools.
+     *
+     *
+     * @param owner
+     *  The owner for which to fetch system purpose attributes
+     *
+     * @return
+     *  a map of attributes belonging to the given owner
+     */
+    public Map<String, Set<String>> getSyspurposeAttributesByOwner(Owner owner) {
+        return this.getSyspurposeAttributesByOwner(owner.getId());
+    }
+
+    /**
+     * Builds a query for fetching the system purpose attributes mapped to the given owner's products
+     *  Will only return attributes for products that are related to unexpired pools.
+     *
+     *
+     * @param ownerId
+     *  The owner ID for which to fetch system purpose attributes
+     *
+     * @return
+     *  a map of attributes belonging to the given owner
+     */
+    public Map<String, Set<String>> getSyspurposeAttributesByOwner(String ownerId) {
+        String sql =
+            "SELECT DISTINCT a.name, a.value  " +
+            "FROM cp2_product_attributes a " +
+            "JOIN cp_pool b ON a.product_uuid=b.product_uuid " +
+            "AND a.name IN ('usage','roles','addons','support_type') " +
+            "AND b.owner_id=:owner_id " +
+            "AND b.endDate > CURRENT_TIMESTAMP ";
+        List<Object[]> result = this.getEntityManager()
+            .createNativeQuery(sql)
+            .setParameter("owner_id", ownerId)
+            .getResultList();
+
+        sql =
+            "SELECT DISTINCT a.name, a.value " +
+            "FROM cp2_product_attributes a " +
+            "JOIN cp_pool b ON a.product_uuid=b.product_uuid " +
+            "AND name = 'support_level' " +
+            "AND b.owner_id=:owner_id " +
+            "AND b.endDate > CURRENT_TIMESTAMP " +
+            "LEFT OUTER join cp2_product_attributes c ON a.product_uuid=c.product_uuid " +
+            "AND c.name='support_level_exempt' " +
+            "WHERE c.value IS NULL or c.value != 'true'";
+
+        result.addAll(this.getEntityManager()
+            .createNativeQuery(sql)
+            .setParameter("owner_id", ownerId)
+            .getResultList());
+
+        Map<String, Set<String>> attributeMap = new HashMap<>();
+        for (Object[] attMap : result) {
+            attributeMap.computeIfAbsent((String) attMap[0], key -> new LinkedHashSet<>())
```

Comment:
I've been thinking about this for a bit, and I'm curious what the intent behind the linked hash set is here. Since we're not doing any sorting in the queries above, retaining the insertion order of the set that gets built here doesn't seem to be very useful -- the output order of the queries aren't guaranteed to be consistent between executions.

If this is something we need to retain, we'll need to add some kind of sorting or some other mechanism to ensure the output of the queries is consistent. Otherwise, we can just use a normal set here and not worry about the order.